modules/common/src/main/scala/com/snowplowanalytics/snowplow/postgres/streaming/KinesisSink.scala

```diff
@@ -0,0 +1,122 @@
+
+import com.snowplowanalytics.snowplow.postgres.config.{StreamSinkConfig, BackoffPolicy}
+
+object KinesisSink {
+
+  lazy val logger = getLogger
+
+  def create[F[_]: Async: Timer](config: StreamSinkConfig.Kinesis, backoffPolicy: BackoffPolicy, blocker: Blocker): Resource[F, StreamSink[F]] =
+    mkProducer(config).map(writeToKinesis(config.streamName, backoffPolicy, blocker))
+
+  private def mkProducer[F[_]: Sync](config: StreamSinkConfig.Kinesis): Resource[F, KinesisProducerClient[F]] =
+    Resource.eval(
+      Sync[F].delay(
+        new KinesisProducerClientImpl[F](Some(mkProducerConfig(config.region.toString)))
+      )
+    )
+
+  private def mkProducerConfig[F[_]](region: String): KinesisProducerConfiguration = {
+    val credentials = new DefaultAWSCredentialsProviderChain()
+    new KinesisProducerConfiguration()
+      .setCredentialsProvider(credentials)
+      .setRegion(region)
+  }
+
+  private def writeToKinesis[F[_]: Async: Timer](streamName: String,
+                                                 backoffPolicy: BackoffPolicy,
+                                                 blocker: Blocker)
+                                                (producer: KinesisProducerClient[F])
+                                                (data: Array[Byte]): F[Unit] = {
+    val res = for {
+      byteBuffer <- Async[F].delay(ByteBuffer.wrap(data))
+      partitionKey <- Async[F].delay(UUID.randomUUID().toString)
+      cb <- producer.putData(streamName, partitionKey, byteBuffer)
+      cbRes <- registerCallback(cb, blocker)
+    } yield cbRes
+    res.retryingOnFailuresAndAllErrors(
+      wasSuccessful = _.isSuccessful,
+      policy = backoffPolicy.retryPolicy,
+      onFailure = (_, retryDetails) => Async[F].delay(logger.warn(s"Failure while writing record to Kinesis - retries so far: ${retryDetails.retriesSoFar}")),
+      onError = (exception, retryDetails) => Async[F].delay(logger.warn(s"Error while writing record to Kinesis - retries so far: ${retryDetails.retriesSoFar} - exception: $exception"))
+    ).void
+  }
+
+  private def registerCallback[F[_]: Async](f: ListenableFuture[UserRecordResult], blocker: Blocker): F[UserRecordResult] =
+    Async[F].async[UserRecordResult] { cb =>
+      Futures.addCallback(
+        f,
+        new FutureCallback[UserRecordResult] {
+          override def onFailure(t: Throwable): Unit = cb(Left(t))
+
+          override def onSuccess(result: UserRecordResult): Unit = cb(Right(result))
+        },
+        (command: Runnable) => blocker.blockingContext.execute(command)
+      )
+    }
+
+  def streamExists[F[_]: Async](config: StreamSinkConfig.Kinesis)(implicit ec: ExecutionContext): F[Boolean] =
+    makeKinesisClient(config.region).use { client =>
+      Async[F].async[Boolean] { cb =>
+        val describeStreamRequest = DescribeStreamRequest.builder()
+          .streamName(config.streamName)
+          .build()
+        client.describeStream(describeStreamRequest)
+          .asScala
```

Comment:
Is it `CompletableFuture`? I've been using this function: https://github.com/typelevel/cats-effect/issues/160#issuecomment-509218922. Comments also explain what's wrong with converting it to Scala `Future`. Not that it's important for this use case, but could be a good educational meterial.

Reply:
Yep, it is `CompletableFuture`. As you said, I guess it is okay to use it like that in here since it is wrapped on `Async[F].async` but thanks for the information, I will use it next time 👍

Reply:
I see. I guess I have still lots to learn 😄  I just skimmed the thread but will read it in detailed definitely. Thanks again.